src/DotVVM.Framework/Controls/DotvvmBindableObject.cs

```diff
@@ -132,17 +132,16 @@ public void MagicSetValue(DotvvmProperty[] keys, object[] values, int hashSeed)
         /// </summary>
         public virtual void SetValue(DotvvmProperty property, object? value)
         {
-            var originalValue = GetValueRaw(property, false);
-            // TODO: really do we want to update the value binding only if it's not a binding
-            if (originalValue is IUpdatableValueBinding && !(value is BindingExpression))
-            {
-                // if the property contains a binding and we are not passing another binding, update the value
-                ((IUpdatableValueBinding)originalValue).UpdateSource(value, this);
-            }
-            else
-            {
-                SetValueRaw(property, value);
-            }
+            SetValueRaw(property, value);
+        }
+
+        /// <summary> Sets the value of specified property by updating the view model this property is bound to. Throws if the property does not contain binding </summary>
+        public void SetValueToSource(DotvvmProperty property, object? value)
+        {
+            var binding = GetBinding(property);
+            if (binding is null)
+                throw new DotvvmControlException(this, $"Property {property} does not contain binding, so it's source can not be updated.");
+            ((IUpdatableValueBinding)binding).UpdateSource(value, this);
```

Comment:
Does the GetBinding always return IUpdatableValueBinding? Otherwise, this can be a source of InvalidCastException